File: Sorting Algorithms/Quick Sort Algorithm/Java/QuickSort.java
/* ===== ===== =====

Theory of Programming

Quick Sort Algorithm
http://theoryofprogramming.com/2015/01/20/quick-sort-algorithm/
GitHub - https://github.com/VamsiSangam/theoryofprogramming
Code Contributor - Aditya Aggarwal (github.com/the-cybersapien, cybersapien.xyz)

===== ===== ===== */

import java.util.Scanner;

public class QuickSort{

	public static void main(String[] args) {
        //Initialize the Scanner Object to take input
        Scanner in = new Scanner(System.in);
        //Get the Size of the Array
        System.out.println("Enter the size of the array:");
        int size = in.nextInt();
        //Initialize the Array for sorting the elements
        int[] array = new int[size];
        //Get the data from user
        System.out.println("Enter the elements of the array: ");

        for (int i = 0; i < size; i++) {
            array[i] = in.nextInt();
        }

        //Call the quickSort function to sort the Array
        quickSort(array, 0, size - 1);

        //Print the Sorted Array
        System.out.println("Sorted Array: ");

        for (int a : array) {
            System.out.print(a + " ");
        }

        System.out.println();
    }

    public static void quickSort(int[] arr, int low, int high) {
        //Check if low is smaller than high
        //This is the exit condition for the recursive solution
        if (low < high){
        	//partition and swap the array
            int partitionIndex = partition(arr, low, high);
            //Quicksort the left of the partition
            quickSort(arr, low, partitionIndex - 1);
            //Quicksort the right of the partition
            quickSort(arr, partitionIndex + 1, high);
        }
    }

    private static int partition(int arr[], int low, int high){
        //Select the pivot as the highest element in the array
        int pivot = arr[high];
        int i = low - 1;

        for (int j = low; j < high; j++) {
            if (arr[j] <= pivot){
                i++;
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }

        int temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;

        return i + 1;
    }
}

File: Sorting Algorithms/Quick Sort Algorithm/Python/QuickSort.py
# ===== ===== =====
#
# Theory of Programming
#
# Quick Sort Algorithm
# http://theoryofprogramming.com/2015/01/20/quick-sort-algorithm/
# GitHub - https://github.com/VamsiSangam/theoryofprogramming
# Code Contributor - Saurabh Rane (github.com/ranesr)
#
# ===== ===== ===== */

# Quick Sort Function
def quickSort(alist):
	quickSortHelper(alist, 0, len(alist) - 1)


# Helper Function
def quickSortHelper(alist, start, end):

	# If there are two or more elements
	if start < end:
		
		# Partition the list
		split = partition(alist, start, end)
		
		# Sort both halves
		quickSortHelper(alist, start, split - 1)
		quickSortHelper(alist, split + 1, end)


# Partition Function
def partition(alist, start, end):

	# Partition around the last value
	pivot = alist[end]
	bottom = start - 1
	top = end
	
	done = 0
	
	# Until all elements are partitioned
	while not done:
		
		# Until we find an out of place element
		while not done:
			# Move 'bottom' up
			bottom += 1
			
			# If we hit the 'top', we are done
			if bottom == top:
				done = 1
				break
			
			# 'bottom' out of place, put it at the 'top', and start searching from the 'top'
			if alist[bottom] > pivot:
				alist[top] = alist[bottom]
				break

		# Until we find an out of place element
		while not done:
			# Move 'top' down
			top -= 1
			
			# If we hit the 'bottom', we are done
			if top == bottom:
				done = 1
				break
				
			# 'top' out of place, put it at the 'bottom', and start searching from the 'bottom'
			if alist[top] < pivot:
				alist[bottom] = alist[top]
				break
	
	# Put the pivot in its place
	alist[top] = pivot
	
	# Return the split point
	return top
	
	
# Example	
alist = [1, 5, 2, 7, 3, 9, 4]
quickSort(alist)
print(alist)
